### PR TITLE
SOC-6231: Implement space templates feature (third iteration)

### PR DESCRIPTION
**Pull request by wkhessairi**

### CODE REVIEW

File: component/service/src/main/java/org/exoplatform/social/rest/impl/spacetemplates/SpaceTemplatesRestResourcesV1.java [hunk 1]
```diff
@@ -58,4 +65,35 @@ public SpaceTemplatesRestResourcesV1(SpaceTemplateService spaceTemplateService)
...
+          @ApiResponse (code = 404, message = "Resource not found"),
+          @ApiResponse (code = 500, message = "Internal server error")})
+  @Produces("image/png")
+  public Response getBannerStream(@Context UriInfo uriInfo,
+                                  @ApiParam(value = "Space template name", required = true) @QueryParam("templateName") String templateName) {
+    SpaceTemplate spaceTemplate = spaceTemplateService.getSpaceTemplateByName(templateName);
+    String bannerPath = spaceTemplate.getBannerPath();
+    if (StringUtils.isNotBlank(bannerPath)) {
+      try {
+        InputStream bannerStream = configurationManager.getInputStream(bannerPath);
```
Comment (boubaker) [inline component/service/src/main/java/org/exoplatform/social/rest/impl/spacetemplates/SpaceTemplatesRestResourcesV1.java]: please change this to
```java
      try (InputStream bannerStream = configurationManager.getInputStream(bannerPath)) {
```

File: component/service/src/main/java/org/exoplatform/social/rest/impl/spacetemplates/SpaceTemplatesRestResourcesV1.java [hunk 2]
```diff
@@ -58,4 +65,35 @@ public SpaceTemplatesRestResourcesV1(SpaceTemplateService spaceTemplateService)
...
+        byte[] bannerByteArray = IOUtils.toByteArray(bannerStream);
+        return Response.ok(bannerByteArray, "image/png").build();
```
Comment (boubaker) [inline component/service/src/main/java/org/exoplatform/social/rest/impl/spacetemplates/SpaceTemplatesRestResourcesV1.java]: You can return directly an input stream inside `Response.ok`.
In addition, could you please add cache management with `EntityTag` like made [here](https://github.com/exodev/social/blob/develop/component/service/src/main/java/org/exoplatform/social/rest/impl/identity/IdentityRestResourcesV1.java#L205)

File: component/service/src/main/java/org/exoplatform/social/rest/impl/spacetemplates/SpaceTemplatesRestResourcesV1.java (hunk 1 above)
Comment (wkhessairi) [reply component/service/src/main/java/org/exoplatform/social/rest/impl/spacetemplates/SpaceTemplatesRestResourcesV1.java]: ok

File: component/service/src/main/java/org/exoplatform/social/rest/impl/spacetemplates/SpaceTemplatesRestResourcesV1.java (hunk 2 above)
Comment (wkhessairi) [reply component/service/src/main/java/org/exoplatform/social/rest/impl/spacetemplates/SpaceTemplatesRestResourcesV1.java]: ok

Comment (wkhessairi) [reply component/service/src/main/java/org/exoplatform/social/rest/impl/spacetemplates/SpaceTemplatesRestResourcesV1.java]: I thnik we cannot use EntityTag, because the banner returned is not set to a specific space to get its lastUpdated time on EntityTag for example. We do not have an attribute on which we can create the EntityTag (even the banner name and path are always the same). WDYT ?

Comment (boubaker) [reply component/service/src/main/java/org/exoplatform/social/rest/impl/spacetemplates/SpaceTemplatesRestResourcesV1.java]: Yep, I waited that you ask ;)
In fact, you can put always the same modification date for now, so that the image are retrieved only once from sources. In the second version of this feature, we will have to change this to use the last updated date from database.

File: component/service/src/main/java/org/exoplatform/social/rest/impl/spacetemplates/SpaceTemplatesRestResourcesV1.java (hunk 1 above)
Comment (boubaker) [reply component/service/src/main/java/org/exoplatform/social/rest/impl/spacetemplates/SpaceTemplatesRestResourcesV1.java]: Not applied, in fact this change is to close the input stream once copied. See [this doc link](https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html)

File: component/service/src/main/java/org/exoplatform/social/rest/impl/spacetemplates/SpaceTemplatesRestResourcesV1.java [hunk 3]
```diff
@@ -55,7 +77,89 @@ public SpaceTemplatesRestResourcesV1(SpaceTemplateService spaceTemplateService)
...
+          notes = "This returns space template banner input stream")
+  @ApiResponses(value = {
+          @ApiResponse (code = 200, message = "Request fulfilled"),
+          @ApiResponse (code = 404, message = "Resource not found"),
+          @ApiResponse (code = 500, message = "Internal server error")})
+  public Response getBannerStream(@Context UriInfo uriInfo,
+                                  @Context Request request,
+                                  @ApiParam(value = "Space template name", required = true) @QueryParam("templateName") String templateName) {
+    SpaceTemplate spaceTemplate = spaceTemplateService.getSpaceTemplateByName(templateName);
+    String bannerPath = spaceTemplate.getBannerPath();
```
Comment (boubaker) [inline component/service/src/main/java/org/exoplatform/social/rest/impl/spacetemplates/SpaceTemplatesRestResourcesV1.java]: Please add a test on `spaceTemplate` nullability, if null return 404 HTTP Response without logging anything or in debug mode at most.

File: component/service/src/main/java/org/exoplatform/social/rest/impl/spacetemplates/SpaceTemplatesRestResourcesV1.java [hunk 4]
```diff
@@ -55,7 +77,89 @@ public SpaceTemplatesRestResourcesV1(SpaceTemplateService spaceTemplateService)
...
+      cc.setMaxAge(86400);
+      builder.cacheControl(cc);
+      return builder.cacheControl(cc).build();
+    }
+    return EntityBuilder.getResponse(new ErrorResource("image does not exist in path: " + bannerPath, "banner not found"),
+        uriInfo, RestUtils.getJsonMediaType(), Response.Status.NOT_FOUND);
+  }
+
+  private List<SpaceTemplate> getPermissionsLabels(List<SpaceTemplate> list, Locale locale) {
+    List<SpaceTemplate> templatelist = new ArrayList<>();
```
Comment (boubaker) [inline component/service/src/main/java/org/exoplatform/social/rest/impl/spacetemplates/SpaceTemplatesRestResourcesV1.java]: We should retrieve permissions for templates only for administrators

File: component/service/src/main/java/org/exoplatform/social/rest/impl/spacetemplates/SpaceTemplatesRestResourcesV1.java [hunk 5]
```diff
@@ -55,7 +77,89 @@ public SpaceTemplatesRestResourcesV1(SpaceTemplateService spaceTemplateService)
...
+            membership = resourceBundle.getString("social.spaces.templates.any");
+          } else {
+            membership = membership.substring(0, 1).toUpperCase() + membership.substring(1);
+          }
+          GroupHandler groupHandler1 = organizationService.getGroupHandler();
+          String groupName = groupId;
+          try {
+            groupName = groupHandler1.findGroupById(groupId).getGroupName();
+          } catch (Exception e) {
+            LOG.error("Cannot find group name, using group id instead", e);
```
Comment (boubaker) [inline component/service/src/main/java/org/exoplatform/social/rest/impl/spacetemplates/SpaceTemplatesRestResourcesV1.java]: When this kind of errors happens ? I think that Group label should be used instead, if not found, the group name will be sufficient.

File: component/service/src/main/java/org/exoplatform/social/rest/impl/spacetemplates/SpaceTemplatesRestResourcesV1.java (hunk 1 above)
Comment (wkhessairi) [reply component/service/src/main/java/org/exoplatform/social/rest/impl/spacetemplates/SpaceTemplatesRestResourcesV1.java]: when applying it, I have a null entity in the ResponseBuilder :) . Maybe we add a finally block ?

File: component/webui/src/main/java/org/exoplatform/social/webui/space/UISpaceSettings.java [hunk 6]
```diff
@@ -113,8 +123,28 @@ private void initTypeSelectBox(UIFormSelectBox typeSelectBox) {
...
-    typeSelectBox.setOptions(templates.stream().sorted(Comparator.comparing(SelectItemOption::getValue)).collect(Collectors.toList()));
-    typeSelectBox.setValue(defaultSpaceTemplate);
+    typeSelectBox.setOptions(templates);
+    SpaceTemplate defaultTemplate = spaceTemplateService.getSpaceTemplateByName(defaultSpaceTemplate);
+    if (defaultTemplate != null && canView(identity, defaultTemplate)) {
+      typeSelectBox.setValue(defaultSpaceTemplate);
+    }
+  }
+
+  private boolean canView(Identity identity, SpaceTemplate template) {
```
Comment (boubaker) [inline component/webui/src/main/java/org/exoplatform/social/webui/space/UISpaceSettings.java]: Please add this logic in `getAllTemplates` method so that we don't have to make this check in UI layer.
In other terms, please add in API of SpaceTemplateService, the method:
```java
public List getAllTemplates(String userId);
```

File: webapp/vue-portlet/src/main/webapp/space-templates-app/spaceTemplatesServices.js [hunk 7]
```diff
@@ -1,5 +1,7 @@
 import { spacesConstants } from '../js/spacesConstants.js';
 
 export function getTemplates(){
-  return fetch(`${spacesConstants.SPACES_TEMPLATES_API}/templates`, {credentials: 'include'}).then(resp => resp.json());
+  // getting language of the PLF
+  const lang = typeof eXo !== 'undefined' ? eXo.env.portal.language : 'en';
```
Comment (boubaker) [inline webapp/vue-portlet/src/main/webapp/space-templates-app/spaceTemplatesServices.js]: You can write it like this:
```javascript
const lang = (window && window.eXo && window.eXo.env && window.eXo.env.portal && window.eXo.env.portal.language) || 'en';
```
This way, we are sure to not have any NPE and that the default value is applied if no value inside `window.eXo.env.portal.language`

File: webapp/vue-portlet/src/main/webapp/space-templates-app/components/ExoSpacesTemplatesSpaces.vue [hunk 8]
```diff
@@ -2,36 +2,53 @@
...
+          {{ $t('social.spaces.templates.banner') }}
+        </th>
+        <th width="20%">
+          {{ $t('social.spaces.templates.permissions') }}
         </th>
       </tr>
       <tr v-if="templates.length === 0">
         <td class="empty center" colspan="12"> {{ $t('social.spaces.templates.noTemplates') }} </td>
       </tr>
       <tr v-for="template in templates" :key="template.name">
```
Comment (boubaker) [inline webapp/vue-portlet/src/main/webapp/space-templates-app/components/ExoSpacesTemplatesSpaces.vue]: You should add v-else here

File: component/service/src/main/java/org/exoplatform/social/rest/impl/spacetemplates/SpaceTemplatesRestResourcesV1.java [hunk 9]
```diff
@@ -55,7 +77,89 @@ public SpaceTemplatesRestResourcesV1(SpaceTemplateService spaceTemplateService)
...
+          @ApiResponse (code = 200, message = "Request fulfilled"),
+          @ApiResponse (code = 404, message = "Resource not found"),
+          @ApiResponse (code = 500, message = "Internal server error")})
+  public Response getBannerStream(@Context UriInfo uriInfo,
+                                  @Context Request request,
+                                  @ApiParam(value = "Space template name", required = true) @QueryParam("templateName") String templateName) {
+    SpaceTemplate spaceTemplate = spaceTemplateService.getSpaceTemplateByName(templateName);
+    String bannerPath = spaceTemplate.getBannerPath();
+    if (StringUtils.isNotBlank(bannerPath)) {
+      EntityTag eTag = new EntityTag(Integer.toString(templateName.hashCode()));
```
[the remark this answers is not in the document]
Comment (wkhessairi) [reply component/service/src/main/java/org/exoplatform/social/rest/impl/spacetemplates/SpaceTemplatesRestResourcesV1.java]: ok

File: component/service/src/main/java/org/exoplatform/social/rest/impl/spacetemplates/SpaceTemplatesRestResourcesV1.java [hunk 10]
```diff
@@ -55,7 +77,89 @@ public SpaceTemplatesRestResourcesV1(SpaceTemplateService spaceTemplateService)
...
+      Response.ResponseBuilder builder = request.evaluatePreconditions(eTag);
+      if (builder == null) {
+        try {
+          InputStream bannerStream = configurationManager.getInputStream(bannerPath);
```
[the remark this answers is not in the document]
Comment (wkhessairi) [reply component/service/src/main/java/org/exoplatform/social/rest/impl/spacetemplates/SpaceTemplatesRestResourcesV1.java]: when applying it, I have a null entity in the ResponseBuilder :) . Maybe we add a finally block ?

File: component/service/src/main/java/org/exoplatform/social/rest/impl/spacetemplates/SpaceTemplatesRestResourcesV1.java (hunk 3 above)
Comment (wkhessairi) [reply component/service/src/main/java/org/exoplatform/social/rest/impl/spacetemplates/SpaceTemplatesRestResourcesV1.java]: ok

File: component/service/src/main/java/org/exoplatform/social/rest/impl/spacetemplates/SpaceTemplatesRestResourcesV1.java (hunk 5 above)
Comment (wkhessairi) [reply component/service/src/main/java/org/exoplatform/social/rest/impl/spacetemplates/SpaceTemplatesRestResourcesV1.java]: user can enter a non valid group id in the permission configuration, so it can not be found. In this case I use the groupName as the user set it.

File: component/webui/src/main/java/org/exoplatform/social/webui/space/UISpaceSettings.java (hunk 6 above)
Comment (wkhessairi) [reply component/webui/src/main/java/org/exoplatform/social/webui/space/UISpaceSettings.java]: ok

File: webapp/vue-portlet/src/main/webapp/space-templates-app/components/ExoSpacesTemplatesSpaces.vue (hunk 8 above)
Comment (wkhessairi) [reply webapp/vue-portlet/src/main/webapp/space-templates-app/components/ExoSpacesTemplatesSpaces.vue]: ok

File: webapp/vue-portlet/src/main/webapp/space-templates-app/spaceTemplatesServices.js (hunk 7 above)
Comment (wkhessairi) [reply webapp/vue-portlet/src/main/webapp/space-templates-app/spaceTemplatesServices.js]: ok

File: component/service/src/main/java/org/exoplatform/social/rest/impl/spacetemplates/SpaceTemplatesRestResourcesV1.java (hunk 4 above)
Comment (wkhessairi) [reply component/service/src/main/java/org/exoplatform/social/rest/impl/spacetemplates/SpaceTemplatesRestResourcesV1.java]: ok

File: component/service/src/main/java/org/exoplatform/social/rest/impl/spacetemplates/SpaceTemplatesRestResourcesV1.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (wkhessairi) [reply component/service/src/main/java/org/exoplatform/social/rest/impl/spacetemplates/SpaceTemplatesRestResourcesV1.java]: done

File: component/service/src/main/java/org/exoplatform/social/rest/impl/spacetemplates/SpaceTemplatesRestResourcesV1.java (hunk 1 above)
Comment (boubaker) [reply component/service/src/main/java/org/exoplatform/social/rest/impl/spacetemplates/SpaceTemplatesRestResourcesV1.java]: Ok, I see, no need to close it. It will be closed here `org.exoplatform.services.rest.impl.provider.InputStreamEntityProvider.writeTo` anyway

File: component/core/src/main/java/org/exoplatform/social/core/space/impl/SpaceTemplateServiceImpl.java [hunk 11]
```diff
@@ -59,7 +70,81 @@ public SpaceTemplateServiceImpl(InitParams params) {
...
+                membership = membership.substring(0, 1).toUpperCase() + membership.substring(1);
+              }
+              GroupHandler groupHandler1 = organizationService.getGroupHandler();
+              String groupName = groupId;
+              Group group = groupHandler1.findGroupById(groupId);
+              if (group != null) {
+                groupName = group.getGroupName();
+              }
+              // Uppercase the first char
+              groupName = groupName.substring(0, 1).toUpperCase() + groupName.substring(1);
```
Comment (boubaker) [inline component/core/src/main/java/org/exoplatform/social/core/space/impl/SpaceTemplateServiceImpl.java]: FYI, you can use StringUtils.capitalize

File: component/core/src/main/java/org/exoplatform/social/core/space/impl/SpaceTemplateServiceImpl.java [hunk 12]
```diff
@@ -59,7 +70,81 @@ public SpaceTemplateServiceImpl(InitParams params) {
...
+          for (String perm : permissions) {
+            try {
+              UserACL.Permission permission = new UserACL.Permission();
+              permission.setPermissionExpression(perm);
+              String groupId = permission.getGroupId();
+              String membership = permission.getMembership();
+              if ("*".equals(membership)) {
+                membership = resourceBundle.getString("social.spaces.templates.any");
+              } else {
+                membership = membership.substring(0, 1).toUpperCase() + membership.substring(1);
```
Comment (boubaker) [inline component/core/src/main/java/org/exoplatform/social/core/space/impl/SpaceTemplateServiceImpl.java]: FYI, you can use StringUtils.capitalize

File: component/core/src/main/java/org/exoplatform/social/core/space/impl/SpaceTemplateServiceImpl.java [hunk 13]
```diff
@@ -59,7 +70,81 @@ public SpaceTemplateServiceImpl(InitParams params) {
...
+            } catch (Exception e) {
+              LOG.warn("Cannot get labelled permission for entry {}", perm, e);
+            }
+          }
+          int index = templatePermission.lastIndexOf(", ");
+          templatePermission.deleteCharAt(index);
+        }
+      }
+      SpaceTemplate editedTemplate = spaceTemplate.clone();
+      editedTemplate.setPermissions(templatePermission.toString());
```
Comment (boubaker) [inline component/core/src/main/java/org/exoplatform/social/core/space/impl/SpaceTemplateServiceImpl.java]: Use editedTemplate.setPermissions labels.
The labels should be added in a dedicated DTO field.
In fact, we will need the original list of technical permissions expressions on UI.

File: component/core/src/main/java/org/exoplatform/social/core/space/impl/SpaceTemplateServiceImpl.java [hunk 14]
```diff
@@ -182,4 +267,24 @@ public void start() {
...
+  private Identity getIdentity(String userId) {
+    IdentityRegistry identityRegistry = CommonsUtils.getService(IdentityRegistry.class);
+    OrganizationService organizationService = CommonsUtils.getService(OrganizationService.class);
+    Identity identity = identityRegistry.getIdentity(userId);
+    if (identity == null) {
+      Collection<Membership> memberships = new ArrayList<>();
+      try {
+        memberships = organizationService.getMembershipHandler().findMembershipsByUser(userId);
+      } catch (Exception e) {
+        LOG.error("Can't get user '{}' memberships", userId, e);
```
Comment (boubaker) [inline component/core/src/main/java/org/exoplatform/social/core/space/impl/SpaceTemplateServiceImpl.java]: We should throw this exception and manage it in UI to display an error

File: component/core/src/main/java/org/exoplatform/social/core/space/impl/SpaceTemplateServiceImpl.java [hunk 15]
```diff
@@ -59,7 +70,81 @@ public SpaceTemplateServiceImpl(InitParams params) {
...
+              if (group != null) {
+                groupName = group.getGroupName();
+              }
+              // Uppercase the first char
+              groupName = groupName.substring(0, 1).toUpperCase() + groupName.substring(1);
+              String key = resourceBundle.getString("social.spaces.templates.permissionEntry");
+              templatePermission.append(key.replace("{0}", membership).replace("{1}", groupName));
+              templatePermission.append(", ");
+            } catch (Exception e) {
+              LOG.warn("Cannot get labelled permission for entry {}", perm, e);
```
Comment (boubaker) [inline component/core/src/main/java/org/exoplatform/social/core/space/impl/SpaceTemplateServiceImpl.java]: We should throw this exception and manage it in UI to display an error

File: component/core/src/main/java/org/exoplatform/social/core/space/spi/SpaceTemplateService.java [hunk 16]
```diff
@@ -38,6 +39,21 @@
    */
   List<SpaceTemplate> getSpaceTemplates();
 
+  /**
+   * Gets all space templates on which the user has permissions
+   *
+   * @param userId
+   */
+  List<SpaceTemplate> getSpaceTemplates(String userId);
```
Comment (boubaker) [inline component/core/src/main/java/org/exoplatform/social/core/space/spi/SpaceTemplateService.java]: Change this to add "throws Exception"

File: component/core/src/main/java/org/exoplatform/social/core/space/spi/SpaceTemplateService.java [hunk 17]
```diff
@@ -38,6 +39,21 @@
...
+
+  /**
+   * Gets all space templates on which the user has permissions with permission labels according to lang
+   * If userId is not an administrators, all permissions are empty
+   *  @param userId
+   * @param lang
+   */
+  List<SpaceTemplate> getLabelledSpaceTemplates(String userId, String lang);
```
Comment (boubaker) [inline component/core/src/main/java/org/exoplatform/social/core/space/spi/SpaceTemplateService.java]: Change this to add "throws Exception"

File: component/core/src/test/java/org/exoplatform/social/core/space/spi/SpaceTemplateServiceTest.java [hunk 18]
```diff
@@ -81,7 +102,7 @@ public void testGetSpaceTemplateByName() {
     SpaceTemplate templateFake = spaceTemplateService.getSpaceTemplateByName("fake");
     // Then
     assertEquals(template.getName(), "classic");
-    assertNull(templateFake);
+    assertEquals(templateFake.getName(), "classic");
```
Comment (boubaker) [inline component/core/src/test/java/org/exoplatform/social/core/space/spi/SpaceTemplateServiceTest.java]: add assertNotNull before this test

File: component/service/src/main/java/org/exoplatform/social/rest/impl/spacetemplates/SpaceTemplatesRestResourcesV1.java [hunk 19]
```diff
@@ -55,7 +65,63 @@ public SpaceTemplatesRestResourcesV1(SpaceTemplateService spaceTemplateService)
...
-    return EntityBuilder.getResponse(spaceTemplateService.getSpaceTemplates(), uriInfo, RestUtils.getJsonMediaType(), Response.Status.OK);
+  public Response getAllTemplates(@Context UriInfo uriInfo,
+                                  @ApiParam(value = "User language", required = true) @QueryParam("lang") String lang) {
+    try {
+      Identity identity = ConversationState.getCurrent().getIdentity();
+      String userId = identity.getUserId();
+      List<SpaceTemplate> list = spaceTemplateService.getLabelledSpaceTemplates(userId, lang);
+      return EntityBuilder.getResponse(list, uriInfo, RestUtils.getJsonMediaType(), Response.Status.OK);
+    } catch (Exception e) {
+      LOG.error("Cannot retrieve permissions labels for space templates", e);
```
Comment (boubaker) [inline component/service/src/main/java/org/exoplatform/social/rest/impl/spacetemplates/SpaceTemplatesRestResourcesV1.java]: The message should be more generic because we don't know the real cause of the exception "Cannot get list of templates for user {}, with lang {}":
```java
      LOG.error("Cannot get list of templates for user {}, with lang {}", userId, lang, e);
```

File: component/service/src/main/java/org/exoplatform/social/rest/impl/spacetemplates/SpaceTemplatesRestResourcesV1.java [hunk 20]
```diff
@@ -55,7 +65,63 @@ public SpaceTemplatesRestResourcesV1(SpaceTemplateService spaceTemplateService)
...
+      return EntityBuilder.getResponse(new ErrorResource("permissions could not be extracted for space templates", "space templates permissions not extracted"),
```
Comment (boubaker) [inline component/service/src/main/java/org/exoplatform/social/rest/impl/spacetemplates/SpaceTemplatesRestResourcesV1.java]: The message should be generic here:
```java
      return EntityBuilder.getResponse(new ErrorResource("Error occurred while getting list of space templates", "space templates permissions not extracted"),
```

File: component/core/src/main/java/org/exoplatform/social/core/space/impl/SpaceTemplateServiceImpl.java (hunk 11 above)
Comment (wkhessairi) [reply component/core/src/main/java/org/exoplatform/social/core/space/impl/SpaceTemplateServiceImpl.java]: ok

File: component/core/src/main/java/org/exoplatform/social/core/space/impl/SpaceTemplateServiceImpl.java (hunk 12 above)
Comment (wkhessairi) [reply component/core/src/main/java/org/exoplatform/social/core/space/impl/SpaceTemplateServiceImpl.java]: ok

File: component/core/src/main/java/org/exoplatform/social/core/space/impl/SpaceTemplateServiceImpl.java (hunk 13 above)
Comment (wkhessairi) [reply component/core/src/main/java/org/exoplatform/social/core/space/impl/SpaceTemplateServiceImpl.java]: ok

File: component/core/src/main/java/org/exoplatform/social/core/space/impl/SpaceTemplateServiceImpl.java (hunk 14 above)
Comment (wkhessairi) [reply component/core/src/main/java/org/exoplatform/social/core/space/impl/SpaceTemplateServiceImpl.java]: ok

File: component/core/src/main/java/org/exoplatform/social/core/space/impl/SpaceTemplateServiceImpl.java (hunk 15 above)
Comment (wkhessairi) [reply component/core/src/main/java/org/exoplatform/social/core/space/impl/SpaceTemplateServiceImpl.java]: ok

File: component/core/src/main/java/org/exoplatform/social/core/space/spi/SpaceTemplateService.java (hunk 16 above)
Comment (wkhessairi) [reply component/core/src/main/java/org/exoplatform/social/core/space/spi/SpaceTemplateService.java]: ok

File: component/core/src/main/java/org/exoplatform/social/core/space/spi/SpaceTemplateService.java (hunk 17 above)
Comment (wkhessairi) [reply component/core/src/main/java/org/exoplatform/social/core/space/spi/SpaceTemplateService.java]: ok

File: component/core/src/test/java/org/exoplatform/social/core/space/spi/SpaceTemplateServiceTest.java (hunk 18 above)
Comment (wkhessairi) [reply component/core/src/test/java/org/exoplatform/social/core/space/spi/SpaceTemplateServiceTest.java]: ok

File: component/service/src/main/java/org/exoplatform/social/rest/impl/spacetemplates/SpaceTemplatesRestResourcesV1.java (hunk 19 above)
Comment (wkhessairi) [reply component/service/src/main/java/org/exoplatform/social/rest/impl/spacetemplates/SpaceTemplatesRestResourcesV1.java]: ok

File: component/service/src/main/java/org/exoplatform/social/rest/impl/spacetemplates/SpaceTemplatesRestResourcesV1.java (hunk 20 above)
Comment (wkhessairi) [reply component/service/src/main/java/org/exoplatform/social/rest/impl/spacetemplates/SpaceTemplatesRestResourcesV1.java]: ok

File: component/core/src/main/java/org/exoplatform/social/core/space/impl/SpaceTemplateServiceImpl.java [hunk 21]
```diff
@@ -59,7 +70,77 @@ public SpaceTemplateServiceImpl(InitParams params) {
...
+      if (perms != null) {
+        Pattern pattern = Pattern.compile(";");
+        List<String> permissions = pattern.splitAsStream(perms).collect(Collectors.toList());
+        for (String perm : permissions) {
+          UserACL.Permission permission = new UserACL.Permission();
+          permission.setPermissionExpression(perm);
+          String groupId = permission.getGroupId();
+          String membership = permission.getMembership();
+          if (identity.isMemberOf(groupId, membership)) {
+            list.add(spaceTemplate);
```
Comment (boubaker) [inline component/core/src/main/java/org/exoplatform/social/core/space/impl/SpaceTemplateServiceImpl.java]: We should change stop the loop on permissions if the user is member of one of the permitted permissions
